### PR TITLE
New py-ipfs related functionality

### PR DESCRIPTION
**Pull request by ntninja**
When trying to implement the simple `ipfs block *` API in py-ipfs, I immediately hit several roadblocks because of limitations in py-datastore.

This PR addresses some of these with a couple more in the works:

  * All datastores are now `AsyncResources` (ie: have a `aclose()` method) to allow for clean shutdown of the datastore if it had acquired any resources to perform its function (think database connections, file descriptors, important shutdown tasks) – they are unused as of yet, but the code I'm currently working on already makes use of this facility and there will likely be code (leveldb) in the future using this.
  * Mount is a new datastore adapter that delegates to other datastores based on the longest matching path prefix – this is needed for parsing the [Datastore.Spec](https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#datastorespec) section of the config file into something useful.
  * Added `stat` as new faster way to query item attributes without needed to issue a full `read` request for the item – the symmetry with `ipfs block stat` should be obvious.

### CODE REVIEW

File: datastore/adapter/mount.py [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+				
+				# Recurse into submount
+				stack.append(current)
+				current = next
+			except KeyError:
+				# Move out of empty submount (this assignment drops the last
+				# reference to the submount)
+				current = stack.pop()
+		
+		assert self.mounts == {}
```
Comment (adamjsawicki) [inline datastore/adapter/mount.py:184]: Is it better to have an `AssertionError` here over an explicit `raise Exception`?

Comment (ntninja) [reply datastore/adapter/mount.py:184]: If the code above is correct then there should be no way that this statement isn't true, hence I thought an assertion to be more appropriate. In general raise is better through if there is any chance that the condition isn't true.

**Review by ntninja — COMMENTED** [review]
Upps, turns out I never actually clicked “submit review” on my reply below. Sorry!

**Comment by AliabbasMerchant** [comment]
LGTM

**Pull request merged**